src/de/deserializer.zig

```diff
@@ -249,137 +267,104 @@ pub fn Deserializer(comptime user_dbt: anytype) type {
+            const token = try self.tokens.next();
+            if (token == .array_begin) {
+                var sa = SeqAccess(Self){ .d = self };
+                return try visitor.visitSeq(allocator, De, sa.seqAccess());
             }
 
             return error.InvalidType;
         }
 
         fn deserializeString(self: *Self, allocator: ?std.mem.Allocator, visitor: anytype) Error!@TypeOf(visitor).Value {
-            if (try self.tokens.next()) |token| {
-                if (token == .String) {
-                    const slice = token.String.slice(self.tokens.slice, self.tokens.i - 1);
-
-                    return switch (token.String.escapes) {
-                        .None => try visitor.visitString(allocator, De, slice),
-                        .Some => blk: {
-                            const s = try unescapeString(allocator.?, token.String, slice);
-                            defer allocator.?.free(s);
-
-                            break :blk try visitor.visitString(allocator, De, s);
-                        },
-                    };
-                }
+            const token = try self.tokens.nextAlloc(self.getAllocator(), .alloc_if_needed);
+            defer self.freeToken(token);
+
+            switch (token) {
+                inline .string, .allocated_string => |slice| return try visitor.visitString(
+                    allocator,
+                    De,
+                    slice,
+                ),
+                else => return error.InvalidType,
             }
-
-            return error.InvalidType;
         }
 
         fn deserializeStruct(self: *Self, allocator: ?std.mem.Allocator, visitor: anytype) Error!@TypeOf(visitor).Value {
-            if (try self.tokens.next()) |token| {
-                if (token == .ObjectBegin) {
-                    var s = StructAccess(Self){ .d = self };
-                    return try visitor.visitMap(allocator, De, s.mapAccess());
-                }
+            if (try self.tokens.next() == .object_begin) {
+                var s = StructAccess(Self){ .d = self };
+                return try visitor.visitMap(allocator, De, s.mapAccess());
             }
 
             return error.InvalidType;
         }
 
         fn deserializeUnion(self: *Self, allocator: ?std.mem.Allocator, visitor: anytype) Error!@TypeOf(visitor).Value {
-            const backup = self.tokens;
-
-            if (try self.tokens.next()) |token| {
-                if (token == .String) {
-                    self.tokens = backup;
-                }
-
-                if (token == .String or token == .ObjectBegin) {
-                    var u = Union(Self){ .d = self };
-                    return try visitor.visitUnion(allocator, De, u.unionAccess(), u.variantAccess());
-                }
+            switch (try self.tokens.peekNextTokenType()) {
+                .string => {},
+                .object_begin => try self.skipToken(),
+                else => return error.InvalidType,
```

Comment:
This approach would have a few issues:
- `peekNextTokenType` only returns the type of the next token, not the contents. We'd still have to call next to get the actual data.
- We always error out in the case of an unexpected token, thus the state the parser is left in is irrelevant.
- First peeking and then getting the token afterwards would have large performance implications, because we'd essentially be parsing everything twice.

I think only peeking if necessary makes more sense here, pretty much for the same reason that the best way to check if a file I want to read exists is to just try to open it and handle errors. It's one less operation, doesn't do any harm and yields the same result.

Reply:
> peekNextTokenType only returns the type of the next token, not the contents. We'd still have to call next to get the actual data.

Sorry, I meant peek first and then get the next token with `next` or `nextAlloc` (not skip it).

> First peeking and then getting the token afterwards would have large performance implications, because we'd essentially be parsing everything twice.

I think it'd actually help performance.

AFAIU, `peekNextTokenType` basically just switches on `self.state`, compares it against a char, and then returns a token. That's a lot less work than what `nextAlloc` does (e.g., modifying the stack, fully parsing numbers, allocating integers/strings, etc.).

In other words, peeking first and then calling `next` or `nextAlloc` basically costs us an extra switch statement, and in return we can fail extremely fast in the bad path (e.g., we never have to allocate anything if we encounter an invalid type or whatever).

> We always error out in the case of an unexpected token, thus the state the parser is left in is irrelevant.

It may be relevant depending on how the deserializer's used.

For example, a while back, I recall you trying to deserialize into a union by writing your own DB that first called `deserializer.deserializeMap` and then, if that failed, called `deserializer.deserializeString` (or something like that). However, that approach doesn't work if the deserializer isn't peekable, since its state would've been modified when `deserializeMap` failed.

Making `Deserializer` peekable would enable this kind of use case and would cost us virtually nothing, as explained above.